notes/dns-client-proxy.md

```diff
@@ -0,0 +1,535 @@
+
+Domain Name Length
+
+: length of Domain Name. Zero if there is no Domain Name
+
+Domain Name
+
+: domain name for authentication or resolving IP addresses
+
+SvcParams Length
+
+: length of SvcParams
+
+SvcParams
+
+: Service parameters
+
+Interface Name Length
+
+: length of Interface Name
+
+Interface Name
+
+: name of outgoing interface for transport connections
+
+This option is designed to give control over what level of detail it
+wants to specify. The first 5 flags (U, UA, A, P, and D) give generate
+requirements for properties of DNS transports that are used by the
+client proxy. The U flag specifies no encryption, the UA flag, at
+least unauthenticated encryption and the A flag require authenticated
+encryptions. With the U, UA, and A flags clear, an effort is made to
+reach authenticated encryption, if that fails unauthenticated encryption
+and if that fails, fall back to an unencrypted transport.
+The P and D flags allow the application to require
+a specific authentication mechanism (PKI or DANE). When both flags are set,
+PKI and DANE are required together. If no flags are set, are least one of
+the two has to succeed if authenticated encryption is required.
+
+The next flags provide more detailed control over which transports
+should be used or not. For each of 5 different transports
+(Do53, DoT, DoH with ALPN h2, DoH with ALPN h3, DoQ) there is a flag
+to allow (A53,AT,AH2,AH3,AQ) or disallow (D53,DT,DH2,DH3,DQ) the use of the
+transport. There is space to add more transports later.
+
+To future proof applications, there is a single flag DO, that disallows
+transports that are not explicitly listed. With this flag clear,
+the application allows future transports. With the flag set, the
+application has to explicitly list which transports can be used.
+For example, by setting only DO and AT, the application forces the
+use of DoT.
+
+Finally, an application can specify its own resolvers or rely on the
+resolvers that are know to the proxy. If ADN Length and Addr Length
+are both zero, then the application requests to resolvers known to
+the proxy. [Note: it is unclear at the moment what to do with any
+Service Parameters]
+
+If the application specifies only an authentication-domain-name then the
+proxy is expected to resolve the name to addresses. If only addresses
+are specified then the proxy assumes that no name is known (though
+a PKI certificate may include an address literal in the subjectAltName).
+If both a name and address are specified then the proxy will use the
+specified addresses and use the name for authentication.
+
+To simplify the encoding of the option, an option with addresses will
+have either IPv4 or IPv6 addresses. If the application wants to specify
+both IPv4 and IPv6 address for a certain authentication-domain-name
+then it has to include two options.
+
+When present, Service Parameters specify how to connect. Otherwise it is
+up to the proxy to try various possibilities.
```

Comment:
Service Parameters can also contain IPv4 and IPv6 addresses (hints) are they simply added to the list of addresses to try with the policy?
Service Parameters can also contain an alpn parameter containing a list of alpns. Does that mean that only the specified transports are available on the in the same Service Parameters given addresses and port?
I guess we need to brain storm a bit about this.

Reply:
https://datatracker.ietf.org/doc/html/draft-ietf-add-dnr-07 says: "The service parameters MUST NOT include "ipv4hint" or "ipv6hint" SvcParams as they are superseded by the included IP addresses."

There is something to be said for that. Certainly to keep the protocol simple.

I think ALPNs in Service Parameters are complex. We certainly need to give a summary of how that works.